modules/apps/segments/segments-web/src/main/resources/META-INF/resources/js/ClayCriteriaBuilder.es.js

```diff
@@ -0,0 +1,197 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import ClayCriteriaGroup from './ClayCriteriaGroup.es';
+import ClayButton from './ClayButton.es';
+
+class ClayCriteriaBuilder extends React.Component {
+	constructor(props) {
+		super(props);
+
+		this.state = {
+			editing: false,
+			initialCriteria: this.props.criteria
+		};
+	}
+
+	render() {
+		const {conjunctions, criteria, operators, properties} = this.props;
+
+		const {editing} = this.state;
+
+		return (
+			<div className="criteria-builder">
+				<div className="criteria-builder-toolbar">
+					<ClayButton
+						label={Liferay.Language.get('edit')}
+						onClick={this._handleToggleEdit}
+					/>
+				</div>
+
+				{this._isCriteriaEmpty() ? (
+					<ClayCriteriaGroup
+						conjunctions={conjunctions}
+						criteria={criteria}
+						criteriaTypes={ClayCriteriaBuilder._buildCriteriaTypes(
+							operators
+						)}
+						editing={editing}
+						onChange={this._updateCriteria}
+						operators={operators}
+						properties={properties}
+						root
+					/>
+				) : (
+					<div
+						className="empty-state"
+						onClick={this._handleNewCriteria}
+					>
+						{Liferay.Language.get('click-to-start-editing')}
+					</div>
+				)}
+			</div>
+		);
+	}
+
+	static _buildCriteriaTypes(operators) {
+		return operators.reduce(
+			(criteriaTypes, {supportedTypes}) => {
+				supportedTypes.forEach(
+					type => {
+						if (!criteriaTypes[type]) {
+							criteriaTypes[type] = operators.filter(
+								operator =>
+									operator.supportedTypes.includes(type)
+							);
+						}
+					}
+				);
+
+				return criteriaTypes;
+			},
+			new Map()
+		);
+	}
+
+	_cleanCriteria(criterion) {
+		const test = criterion
+			.filter(
+				({items}) => {
+					return items ? items.length : true;
+				}
+			)
+			.map(
+				item => {
+					return item.items ?
+						Object.assign(
```

Comment:
You could also use an object spread here too. You might want to just do a full search and try to use either assign or spread everywhere. Its usually best to pick one way of doing it and try to stick with it for consistency in readability. Using either one is fine, but I'd just try to stick with it.

Using a spread might look better with the csf rules and the amount of indentation, but I'm not positive on that.

Reply:
I don't think it'll help with indentation since it mainly only removes `Object.assign`. But it does help with readability without `Object.assign`. The spread could seem a bit odd but I'll still go ahead with the change.

Example of what the change will look like:

```javascript
.map(
	item => {
		return item.items ?
			{
				...item,
				...{
					items: this._cleanCriteria(item.items)
				}
			} :
			item;
	}
);